### PR TITLE
rfctr cuisine nd csne builder to use apps

### PR DESCRIPTION
**Pull request by abdulrahmantkhalifa**
skydns module added + builder turned to apps

added caddy module + refactored code that uses it

stor module added

added cuisine syncthing module , no ays deps on it

added redis module to cusine+refactoring  to it

cuisine added mongo module + refctring for it

nourhan added etcd + fs to cuisine module

added controller module to cuisine

added grafana cuisine mmodule and some bug fixes

added influx module to cuisine + refactoring

-grafan pep8 formating

added core module to cuisine

cluster in etcd cuisine added to module

refactoring cuisine to use new core hook

moved file from cuisine_ to cuisine

added apps to folder

renamed cuisine core (agent module) to coreos

original code in  https://github.com/Jumpscale/jumpscale_core8/tree/cuisine_refactoring

### CODE REVIEW

File: lib/JumpScale/tools/cuisine/apps/CuisineApps.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    def vulcand(self):
+        if self._vulcand is None:
+            self._vulcand = Vulcand(self.executor, self.cuisine)
+        return self._vulcand
+
+    @property
+    def weave(self):
+        if self._weave is None:
+            self._weave = Weave(self.executor, self.cuisine)
+        return self._
```
Comment (rkhamis) [inline lib/JumpScale/tools/cuisine/apps/CuisineApps.py]: maybe you meant to return self._weave?

Comment (zaibon) [inline lib/JumpScale/tools/cuisine/apps/CuisineApps.py]: Typo here

File: lib/JumpScale/tools/cuisine/apps/CuisineApps.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        if len(error_files) == 0:
+            print("all uploaded ok")
+        else:
+            raise RuntimeError('some files didnt upload properly. %s' % ("\n".join(error_files)))
+
+        metadataPath = j.sal.fs.joinPaths(output_dir, "md", "%s.flist" % namespace)
+        print('uploading %s' % metadataPath)
+        store_client.putStaticFile(namespace+".flist", metadataPath)
+
+    def caddyConfig(self,sectionname,config):
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/apps/CuisineApps.py]: Need to be move into caddy module

File: lib/JumpScale/tools/cuisine/apps/CuisineFs.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.cuisine.golang.install()
+        self.cuisine.golang.get("github.com/g8os/fs", action=True)
+        self.cuisine.core.dir_ensure("$tmplsDir/cfg/fs")
+        self.cuisine.core.file_copy("$goDir/bin/fs", "$base/bin")
+        self.cuisine.core.file_write("$goDir/src/github.com/g8os/fs/config/config.toml", content)
+        self.cuisine.core.file_copy("$goDir/src/github.com/g8os/fs/config/config.toml", "$tmplsDir/cfg/fs")
+        if start:
+            self.start()
+
+    def start(self):
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/apps/CuisineFs.py]: Here we need to download the js8_opt.flist from https://stor.jumpscale.org/storx/static/js8_opt.flist and put it in the correct location according to the config.

File: lib/JumpScale/tools/cuisine/apps/CuisineMongoCluster.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
+from JumpScale import j
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/apps/CuisineMongoCluster.py]: Can we merge that with the mongo module. so we have al mongo related code in the same place?

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 5]
```diff
@@ -1,1064 +0,0 @@
-
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: We still need the builder module.
The builder module calls the build methods from each app module.
In this module need : 
- the all method to build all apps in one command 
- sandbox related methods

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 6]
```diff
@@ -1,1064 +0,0 @@
...
-mkdir -p $tmplsDir/grafana
-cp -rn conf public vendor $tmplsDir/grafana
-mkdir -p %s
-"""%(logDir)
-
-            C = self.cuisine.bash.replaceEnvironInText(C)
-            C = self.cuisine.core.args_replace(C)
-            self.cuisine.core.run_script(C, profile=True, action=True)
-            self.cuisine.bash.addPath(self.cuisine.core.args_replace("$binDir"), action=True)
-            cfg = self.cuisine.core.file_read("$tmplsDir/grafana/conf/defaults.ini")
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: path to template config is `$tmplsDir/cfg` not just `$tmplsDir`

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 7]
```diff
@@ -1,1064 +0,0 @@
...
-            cfg = cfg.replace('data = data', 'data = %s'%(dataDir))
-            cfg = cfg.replace('logs = data/log', 'logs = %s'%(logDir))
-            self.cuisine.core.file_write("$tmplsDir/grafana/conf/defaults.ini", cfg)
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: path to template config is `$tmplsDir/cfg` not just `$tmpls

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 8]
```diff
@@ -1,1064 +0,0 @@
...
-        if self.cuisine.core.isUbuntu:
-            dataDir = self.cuisine.core.args_replace("$varDir/data/grafana")
-            logDir = '%s/log' %(dataDir)
-            C= """
-cd $tmpDir
-wget https://grafanarel.s3.amazonaws.com/builds/grafana-2.6.0.linux-x64.tar.gz
-tar -xvzf grafana-2.6.0.linux-x64.tar.gz
-cd grafana-2.6.0
-cp bin/grafana-server $binDir
-mkdir -p $tmplsDir/grafana
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: path to template config is $tmplsDir/cfg not just $tmplsDir

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    def __init__(self, executor, cuisine):
+        self.executor = executor
+        self.cuisine = cuisine
+
+    def all(self, start=False, sandbox=False, stor_addr=None):
+            self.cuisine.installerdevelop.pip()
+            self.cuisine.installerdevelop.python()
+            self.cuisine.installerdevelop.jumpscale8()
+            self.mongodb.build(start=start)
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: This is not correct. 
it should be `self.cuisine.app.mongodb.build`
same for all app

File: lib/JumpScale/tools/cuisine/CuisineBuilder.py [hunk 10]
```diff
@@ -18,19 +18,19 @@ def all(self, start=False, sandbox=False, stor_addr=None):
...
-            self.redis.build(start=start, force=True)
-            self.core.build(start=start)
-            self.syncthing.build(start=start)
-            self.controller.build(start=start)
-            self.fs.build(start=start)
-            self.stor.build(start=start)
-            self.etcd.build(start=start)
-            self.caddy.build(start=start)
+            self.cuisine.apps.mongodb.build(start=start)
+            self.cuisine.apps.cuisine.portal.install(start=start)
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/CuisineBuilder.py]: Portal module need to be moved under app.
Same for other module like geodns, dnsmasq ...

File: lib/JumpScale/tools/cuisine/apps/CuisineFs.py [hunk 11]
```diff
@@ -47,11 +47,13 @@ def build(self, start=False):
...
         self.cuisine.golang.install()
-        self.cuisine.golang.get("github.com/g8os/fs", action=True)
+        self.cuisine.golang.godep("github.com/g8os/fs", action=True)
+        self.cuisine.core.run("cd %s && go build ." % "$goDir/src/github.com/g8os/fs", profile=True)
         self.cuisine.core.dir_ensure("$tmplsDir/cfg/fs")
-        self.cuisine.core.file_copy("$goDir/bin/fs", "$base/bin")
+        self.cuisine.core.file_copy("$goDir/src/github.com/g8os/fs/fs", "$base/bin")
         self.cuisine.core.file_write("$goDir/src/github.com/g8os/fs/config/config.toml", content)
         self.cuisine.core.file_copy("$goDir/src/github.com/g8os/fs/config/config.toml", "$tmplsDir/cfg/fs")
+        self.cuisine.core.file_download(" https://stor.jumpscale.org/storx/static/js8_opt.flist", "$tmplsDir/cfg/fs/")
```
Comment (zaibon) [inline lib/JumpScale/tools/cuisine/apps/CuisineFs.py]: Two things here
First, the destination of file_download should be the actual file you want to write, so the line should be
`self.cuisine.core.file_download("https://stor.jumpscale.org/storx/static/js8_opt.flist", "$tmplsDir/cfg/fs/js8_opt.flist")`

Second the execution of this line raise an error :
```
E:curl: (1) Protocol " https" not supported or disabled in libcurl
RuntimeError: Could not download: https://stor.jumpscale.org/storx/static/js8_opt.
```
This is due to the starting space before `https://...`
This me thing this code has never even been tested before added to this pull request.
Can you make sure you actually run your code at least once before considering to add it to the PR. Thanks

**Pull request closed without merging**